src/zql/ivm/graph/operators/filter-operator.test.ts

```diff
@@ -1,98 +1,81 @@
+  const out = input.filter(_ => false);
+  const items: E[] = [];
+  out.effect((e: E) => {
+    items.push(e);
+  });
 
-  inputWriter.queueData([
-    1,
-    new Multiset([
-      [1, 1],
-      [2, 2],
-      [1, -1],
-      [2, -2],
-    ]),
+  input.newData(1, [
+    [{id: 1}, 1],
+    [{id: 2}, 2],
+    [{id: 1}, -1],
+    [{id: 2}, -2],
   ]);
-  inputWriter.notify(1);
-  inputWriter.notifyCommitted(1);
-  const items = outReader.drain(1);
+  input.commit(1);
 
-  expect([...items![1].entries].length).toBe(0);
+  expect(items.length).toBe(0);
 });
 
 test('emits all rows that pass the filter (including deletes / retractions)', () => {
-  const inputWriter = new DifferenceStreamWriter<number>();
-  const inputReader = inputWriter.newReader();
-  const output = new DifferenceStreamWriter<number>();
-
-  new FilterOperator(inputReader, output, _ => true);
+  const input = new DifferenceStream<E>();
+  const out = input.filter(_ => true);
 
-  const outReader = output.newReader();
-  outReader.setOperator(new NoOp());
+  const items: Entry<E>[] = [];
+  out.effect((e: E, mult: number) => {
+    items.push([e, mult]);
+  });
 
-  inputWriter.queueData([
-    1,
-    new Multiset([
-      [1, 1],
-      [2, 2],
-      [1, -1],
-      [2, -2],
-    ]),
+  input.newData(1, [
+    [{id: 1}, 1],
+    [{id: 2}, 2],
+    [{id: 1}, -1],
+    [{id: 2}, -2],
   ]);
-  inputWriter.notify(1);
-  inputWriter.notifyCommitted(1);
-  const items = outReader.drain(1);
+  input.commit(1);
 
-  const entries = [...items![1].entries];
-  expect(entries).toEqual([
-    [1, 1],
-    [2, 2],
-    [1, -1],
-    [2, -2],
+  expect(items).toEqual([
+    [{id: 1}, 1],
+    [{id: 2}, 2],
+    [{id: 1}, -1],
+    [{id: 2}, -2],
   ]);
 });
 
 test('test that filter is lazy / the filter is not actually run until we pull on it', () => {
-  const inputWriter = new DifferenceStreamWriter<number>();
-  const inputReader = inputWriter.newReader();
-  const output = new DifferenceStreamWriter<number>();
-
+  const input = new DifferenceStream<E>();
   let called = false;
-  new FilterOperator(inputReader, output, _ => {
+  const out = input.filter(_ => {
     called = true;
     return true;
   });
+  const msgs: Iterable<Entry<E>>[] = [];
+  out.debug((_: Version, data: Iterable<Entry<E>>) => {
+    msgs.push(data);
+  });
 
-  const outReader = output.newReader();
-  outReader.setOperator(new NoOp());
-
-  inputWriter.queueData([
-    1,
-    new Multiset([
-      [1, 1],
-      [2, 2],
-      [1, -1],
-      [2, -2],
-    ]),
+  input.newData(1, [
+    [{id: 1}, 1],
+    [{id: 2}, 2],
+    [{id: 1}, -1],
+    [{id: 2}, -2],
   ]);
-  inputWriter.notify(1);
-  inputWriter.notifyCommitted(1);
+  input.commit(1);
 
   // we run the graph but the filter is not run until we pull on it
   expect(called).toBe(false);
 
-  const items = outReader.drain(1);
   // consume all the rows
-  [...items![1].entries];
+  for (const m of msgs) {
+    [...m];
```

Comment:
What is this testing?

Are you doing it this way to drain the iterable?

It seems like it would be clearer if you asserted that the array looks like you expect it?